src/data-operations/data-util.ts

```diff
@@ -0,0 +1,179 @@
+        });
+        if (!state.strategy) {
+            return data;
+        }
+        return state.strategy.sort(data, state.expressions);
+    }
+    static page<T> (data: T[], state: PagingState): T[] {
+        if (!state) {
+            return data;
+        }
+        var len = data.length,
+            index = state.index,
+            res = [],
+            recordsPerPage = state.recordsPerPage;
+        state.metadata = {
+            countPages: 0,
+            error: PagingError.None,
+            countRecords: data.length
+        };
+        if (index < 0 || isNaN(index)) {
+            state.metadata.error = PagingError.IncorrectPageIndex;
+            return res;
+        }
+        if (recordsPerPage <= 0 || isNaN(recordsPerPage)) {
+            state.metadata.error = PagingError.IncorrectRecordsPerPage;
+            return res;
+        }
+        state.metadata.countPages = Math.ceil(len / recordsPerPage);
+        if (index >= state.metadata.countPages) {
+            state.metadata.error = PagingError.IncorrectPageIndex;
+            return res;
+        }
+        return data.slice(index * recordsPerPage, (index + 1) * recordsPerPage);
+    }
+    static filter<T> (data: T[],
+                    state: FilteringState): T[] {
+        if (!state || !state.expressions) {
+            return data;
+        }
+        // set defaults
+        state.expressionDefaults = DataUtil.mergeDefaultProperties(state.expressionDefaults, 
+                                        SortingStateDefaults.expressionDefaults);
+        // set defaults
+        DataUtil.mergeDefaultProperties(state, filteringStateDefaults);
+        state.expressions.forEach((expr) => {
+            expr.ignoreCase = expr.ignoreCase === undefined ?
+                            (state.expressionDefaults || {}).ignoreCase: expr.ignoreCase;
+        });
+        if (!state.strategy) {
+            return data;
+        }
+        return state.strategy.filter(data, state.expressions, state.logic);
+    }
+    static process<T> (data: T[], state: DataState): T[] {
+        if (!state) {
+            return data;
+        }
+        if (state.filtering) {
+            data = DataUtil.filter(data, state.filtering);
+        }
+        if (state.sorting) {
+            data = DataUtil.sort(data, state.sorting);
+        }
+        if (state.paging) {
+            data = DataUtil.page(data, state.paging);
+        }
+        return data;
+    }
+    /* CRUD operations */
+    // access data records
+    static getIndexOfRecord (data: any[], record: Object): number {
+        data = data || [];
+        return data.indexOf(record);
+    }
+    static getRecordByIndex (data: any[], index: number) {
+        if (index < 0 || !data) {
+            return undefined;
+        }
+        data = data || [];
+        return data[index];
+    }
+    static getRecordInfoByKeyValue (data: any[], fieldName: string, value: any): {index: number, record: Object} {
```

Comment:
This is the method that will be left, but it will be moved to the `DataContainer` class